### PR TITLE
Correct Vector2 and Rect2 to Rect2i and Vector2i in many cases

### PR DESCRIPTION
**Pull request by AaronRecord**
(can be merged independently from https://github.com/godotengine/godot/pull/52456)

### CODE REVIEW

**Comment by kleonc** [comment]
Any reason for not using `Point2i`/`Size2i`? Personally I don't mind using plain `Vector2i` everywhere but I guess `Point2i`/`Size2i` is preferred in the code base? :thinking: Also in `image.h` you've left some `Point2i`/`Size2i`.

And have you checked the places where the changed methods are called? These probably need to be checked to avoid potential regressions. E.g. if there would be a code like `some_image.get_size() * some_vector2` now it would be using `Vector2i::operator*` instead of `Vector2::operator*` and thus it could behave differently.

**Comment by LightningAA** [comment]
> Any reason for not using `Point2i`/`Size2i`? Personally I don't mind using plain `Vector2i` everywhere but I guess `Point2i`/`Size2i` is preferred in the code base? 🤔 Also in `image.h` you've left some `Point2i`/`Size2i`.

I'm personally in favor of just using `Vector2i` without any extra aliases (I think it just adds unnecessary complexity), but I guess this should probably all be done at once after a consensus has been reached (e.g. why does `get_size()` return `Vector2i` and not `Size2i`?)

> And have you checked the places where the changed methods are called? These probably need to be checked to avoid potential regressions. E.g. if there would be a code like `some_image.get_size() * some_vector2` now it would be using `Vector2i::operator*` instead of `Vector2::operator*` and thus it could behave differently.

Yeah, that occurred to me, but I haven't done any thorough investigation. I'll look into this.

**Comment by LightningAA** [comment]
> And have you checked the places where the changed methods are called? These probably need to be checked to avoid potential regressions. E.g. if there would be a code like `some_image.get_size() * some_vector2` now it would be using `Vector2i::operator*` instead of `Vector2::operator*` and thus it could behave differently.

I've gone through, and I ended going down a rabbit hole of Vector2 vs Vector2i, I converted a bunch of cases, but there's a lot and I only have so much time and energy 😄

I think it'll be fine, I already found several existing cases of Vector2i being used where Vector2 was expected or vice versa. Those cases can be fixed in future PRs.

**Comment by kleonc** [comment]
> I think it'll be fine

CI thinks differently. :smile: As you can see in the failing checks now `some_texture.get_size() * 0.5` would result in implicitly converting `0.5` to `int32_t` and `Vector2i::operator* (int32_t)` would be used, not `Vector2::operator* (real_t)` which would be used in there before the changes. If every such problem could be easily spotted then fixing these wouldn't be a big problem. The real problem/question is if there are unspotted places where introduced changes silently change the current behavior. That's what I was referring in my previous comment. And as you've already experienced it's easy to not catch everything being affected by these changes.

---

Also I see you've changed for example the type of the `AtlasTexture`'s `region` to be `Rect2i` but currently it's fine to use non-integer position for the region, it's being respected when rendering (however region's size seems to be treated as truncated to ints). So it's kinda a regression. In fact, I'd say for an `AtlasTexture` it should be perfectly fine to allow both `region` and `margin` to have non-integer `position` and `size` (and thus the result of `get_size()`? :thinking:). In the end it's just a concept used to refer to a part of a "real texture" and I see no reason to limit it to allow representing only integer-only parts/regions (there's no "physical limitation" like with "real textures" which always have integer-only size). So this probably needs to be discussed before changing it anyhow.

**Comment by LightningAA** [comment]
> CI thinks differently. 😄 As you can see in the failing checks now `some_texture.get_size() * 0.5` would result in implicitly converting `0.5` to `int32_t` and `Vector2i::operator* (int32_t)` would be used, not `Vector2::operator* (real_t)` which would be used in there before the changes. 

Yeah, it should probably be ` / 2`, not ` * 0.5`
> If every such problem could be easily spotted then fixing these wouldn't be a big problem. The real problem/question is if there are unspotted places where introduced changes silently change the current behavior. That's what I was referring in my previous comment. And as you've already experienced it's easy to not catch everything being affected by these changes.

Agreed, but there are already many cases of Vector2/Rect2 being used where Vector2i/Rect2i are expected or vice versa, so it's probably not a big problem. The problem is that a lot of the code is so intertwined, so you either need to update a lot of Rect2/Vector2's to Rect2i/Vector2i's all at once, or you just draw a line at a certain point to limit the scope of the PR.

> Also I see you've changed for example the type of the `AtlasTexture`'s `region` to be `Rect2i` but currently it's fine to use non-integer position for the region, it's being respected when rendering (however region's size seems to be treated as truncated to ints). So it's kinda a regression. In fact, I'd say for an `AtlasTexture` it should be perfectly fine to allow both `region` and `margin` to have non-integer `position` and `size` (and thus the result of `get_size()`? 🤔). In the end it's just a concept used to refer to a part of a "real texture" and I see no reason to limit it to allow representing only integer-only parts/regions (there's no "physical limitation" like with "real textures" which always have integer-only size). So this probably needs to be discussed before changing it anyhow.

I wasn't sure about this, I can revert this change. I guess I hadn't thought about scaling sprites, so maybe it could be useful to allow non integer offsets?

File: core/io/image.cpp [hunk 1]
```diff
@@ -445,8 +445,8 @@ int Image::get_height() const {
 	return height;
 }
 
-Vector2 Image::get_size() const {
-	return Vector2(width, height);
+Vector2i Image::get_size() const {
```
Comment (Chaosus) [inline core/io/image.cpp:448]: I would rename it to `Size2i` for clarity.

**Comment by AaronRecord** [comment]
Superseded by https://github.com/godotengine/godot/pull/62871

**Pull request closed without merging**